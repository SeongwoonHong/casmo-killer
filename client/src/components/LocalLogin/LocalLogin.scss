.Local-login {
    &__form {
        &__btn:last-child {
            margin: $user-box-padding;
            padding: 10px 0;
            width: calc(100% - #{$user-box-padding * 2});
        }
    }
    &__options {
        @include flex-box(space-between);
        padding: 0 $user-box-padding;
        &__links {
            color: $teal-1000;
            font-size: 12px;
        }
    }
}
