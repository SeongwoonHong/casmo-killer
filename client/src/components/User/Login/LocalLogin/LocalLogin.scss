.user-login-local {
    > form {
        > .user-form-button {
            font-size: 16px;
            margin: 15px auto;
            padding: 10px 0;
            width: calc(100% - 30px);
        }
    }
    > .other-options {
        @include flex-box(space-between);
        padding: 0 15px 15px;
        > a {
            color: palette(Teal, 900);
            font-size: 13px;
        }
    }
}
