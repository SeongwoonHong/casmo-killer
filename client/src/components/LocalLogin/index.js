import LocalLogin from './LocalLogin';

export default LocalLogin;
