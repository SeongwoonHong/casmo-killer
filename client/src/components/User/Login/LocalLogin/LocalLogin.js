import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';

import { validateEmail } from 'sharedUtils/inputValidators';

import FormMessage from 'sharedComponents/FormMessage';

import UserPageContainer from '../../shared/UserPageContainer';
import UserInputField from '../../shared/UserInputField';

import './LocalLogin.scss';

const initialState = {
  isLoading: false,
  email: '',
  password: '',
  message: '',
  successMsg: '',
};

class LocalLogin extends Component {

  constructor(props) {
    super(props);
    this.state = initialState;
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.isLogin !== nextProps.isLogin) {
      this.setState(initialState);
    }
  }

  onChangeHandler = (e) => {
    this.setState({ [e.name]: e.value });
  };

  onSubmitHandler = () => {

    const { isLogin } = this.props;

    this.setState({
      isLoading: true,
      message: '',
      successMsg: ''
    });

    if (isLogin) {
      this.onLogin();
    } else {
      this.onRegister();
    }

  };

  async onLogin() {

    const { onSuccess } = this.props;
    const { email, password } = this.state;

    if (email.length === 0 || password.length === 0) {

      this.setState({
        isLoading: false,
        message: 'Please enter your email and password.'
      });

    } else {

      try {

        const { data } = await axios.post('/api/auth/login/local', {
          email, password
        });

        if (data && data.user) {

          onSuccess(data.user);

        } else {

          this.setState({
            isLoading: false,
            message: 'Failed to communicate with the server.'
          });

        }

      } catch (error) {

        console.error(error.response.data.error);
        this.setState({
          isLoading: false,
          message: error.response.data.message
        });

      }

    }

  }

  async onRegister() {

    const { email } = this.state;

    const message = await validateEmail(email);

    if (message.length > 0) {

      this.setState({
        isLoading: false,
        message
      });

    } else {

      try {

        const { data } = await axios.post('/api/auth/request/verification', {
          email
        });

        if (data && data.message) {

          this.setState({
            isLoading: false,
            successMsg: data.message
          });

        } else {

          this.setState({
            isLoading: false,
            message: 'Failed to communicate with the server.'
          });

        }

      } catch (error) {

        console.error(error);
        this.setState({
          isLoading: false,
          message: error.response.data.message
        });

      }

    }

  }

  render() {

    const {
      isLoading, email, password, message, successMsg
    } = this.state;

    const { isLogin, redirectUrl } = this.props;

    const formText = isLogin ? 'Log In' : 'Sign Up';

    return ([
      <UserPageContainer
        key={ 0 }
        className="Local-login"
        formTitle={ `${formText} with Email`}
        isLoading={ isLoading }
        onSubmit={ this.onSubmitHandler }
        button={
          successMsg.length > 0
            ? (
              <Link
                to={ redirectUrl }
                className="user-form-button">
                Go Back
              </Link>
            )
            : (
              <button
                type="submit"
                className="user-form-button">
                { formText }
              </button>
            )
        }>
        <FormMessage message={ message } />
        <UserInputField
          isVisible={ !successMsg.length > 0 }
          type="email"
          name="email"
          onChange={ this.onChangeHandler }
          value={ email } />
        <FormMessage
          message={ successMsg }
          type="success" />
        <UserInputField
          isVisible={ isLogin }
          type="password"
          name="password"
          onChange={ this.onChangeHandler }
          value={ password } />
      </UserPageContainer>,
      <div
        key={ 1 }
        className="other-options">
        <Link to={ `/user/auth/${isLogin ? 'signup' : 'login'}` }>
          {
            isLogin
              ? 'Sign Up For Free'
              : 'Log In With Email'
          }
        </Link>
        <Link to="/user/recover">
          Forgot Your Password?
        </Link>
      </div>
    ]);

  }

}

export default LocalLogin;
