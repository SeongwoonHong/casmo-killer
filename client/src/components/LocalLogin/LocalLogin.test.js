import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { expect } from 'chai';
import { mount, shallow } from 'enzyme';

import LocalLogin from './LocalLogin';

describe('<LocalLogin />', () => {

  let wrapper = null;

  beforeEach(() => {
    wrapper = mount(
      <MemoryRouter>
        <LocalLogin />
      </MemoryRouter>
    );
  });

  it('render two inputs and one submit button', () => {

    const UserPageContainer = wrapper.find('UserPageContainer');
    const UserInputFields = wrapper.find('UserInputField');

    expect(UserInputFields.first().find('input').is('input[type="email"]')).to.be.true;
    expect(UserInputFields.last().find('input').is('input[type="password"]')).to.be.true;
    expect(UserPageContainer.find('button').is('button[type="submit"]')).to.be.true;

  });

});
