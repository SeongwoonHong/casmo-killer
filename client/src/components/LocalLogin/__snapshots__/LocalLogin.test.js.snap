// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocalLogin /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocalLogin />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Array [
    <UserPageContainer
      button={
            <button
                  className="User__form__btn Local-login__form__btn"
                  type="submit"
            >
                  Sign Up
            </button>
      }
      className="Local-login"
      disabled={false}
      formMsg=""
      formTitle="Sign Up with Email"
      icon=""
      isLoading={false}
      onSubmit={[Function]}
      title=""
>
      <FormMessage
            message=""
            type="error"
      />
      <UserInputField
            disabled={false}
            isVisible={true}
            message=""
            name="email"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value=""
      />
      <FormMessage
            message=""
            type="success"
      />
      <UserInputField
            disabled={false}
            isVisible={true}
            message=""
            name="password"
            onChange={[Function]}
            placeholder=""
            title=""
            type=""
            value=""
      />
</UserPageContainer>,
    <div
      className="Local-login__options"
>
      <Link
            className="Local-login__options__links"
            replace={false}
            to="/user/auth/login"
      >
            Log In With Email
      </Link>
      <Link
            className="Local-login__options__links"
            replace={false}
            to="/user/recover"
      >
            Forgot Your Password?
      </Link>
</div>,
  ],
  Symbol(enzyme.__nodes__): Array [
    Array [
      <UserPageContainer
        button={
                <button
                        className="User__form__btn Local-login__form__btn"
                        type="submit"
                >
                        Sign Up
                </button>
        }
        className="Local-login"
        disabled={false}
        formMsg=""
        formTitle="Sign Up with Email"
        icon=""
        isLoading={false}
        onSubmit={[Function]}
        title=""
>
        <FormMessage
                message=""
                type="error"
        />
        <UserInputField
                disabled={false}
                isVisible={true}
                message=""
                name="email"
                onChange={[Function]}
                placeholder=""
                title=""
                type=""
                value=""
        />
        <FormMessage
                message=""
                type="success"
        />
        <UserInputField
                disabled={false}
                isVisible={true}
                message=""
                name="password"
                onChange={[Function]}
                placeholder=""
                title=""
                type=""
                value=""
        />
</UserPageContainer>,
      <div
        className="Local-login__options"
>
        <Link
                className="Local-login__options__links"
                replace={false}
                to="/user/auth/login"
        >
                Log In With Email
        </Link>
        <Link
                className="Local-login__options__links"
                replace={false}
                to="/user/recover"
        >
                Forgot Your Password?
        </Link>
</div>,
    ],
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
